src/WebJobs.Script.WebHost/Management/LinuxInstanceManager.cs

```diff
@@ -62,24 +98,28 @@ public bool StartAssignment(HostAssignmentContext context)
                 return false;
             }
 
-            if (context.IsWarmupRequest)
-            {
-                // Based on profiling download code jit-ing holds up cold start.
-                // Pre-jit to avoid paying the cost later.
-                Task.Run(async () => await DownloadWarmupAsync(context.GetRunFromPkgContext()));
-                return true;
-            }
-            else if (_assignmentContext == null)
+            return true;
+        }
+
+        private async Task AssignAsync(HostAssignmentContext assignmentContext)
+        {
+            if (assignmentContext.IsWarmupRequest)
             {
-                lock (_assignmentLock)
+                try
```

Comment:
Can you please move the logic to handle a warmup request into its own method (e.g., `HandleWarmupRequestAsync` or something similar).